File: src/lib/progression.cpp
/*
Copyright 2016, Blur Studio

This file is part of Simplex.

Simplex is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

Simplex is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with Simplex.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "shapeBase.h"
#include "progression.h"
#include "simplex.h"
#include "enums.h"

#include "rapidjson/document.h"
#include "rapidjson/rapidjson.h"

#include <algorithm>  // for sort
#include <utility>
#include <vector>
#include <string>

using namespace simplex;
class simplex::Shape;

Progression::Progression(const std::string &name, const ProgPairs &pairs, ProgType interp):
		ShapeBase(name), pairs(pairs), interp(interp) {
	std::sort(this->pairs.begin(), this->pairs.end(),
		[](const ProgPair &a, const ProgPair &b) {
			return a.second < b.second;
		}
	);
}

size_t Progression::getInterval(double tVal, const std::vector<double> &times, bool &outside){
	if (times.size() <= 1){
		outside = true;
		return 0;
	}
	outside = tVal < times[0] || tVal > times[times.size() - 1];
	if (tVal >= times[times.size() - 2]){
		return times.size() - 2;
	}
	else if (tVal < times[0]){
		return 0;
	}
	else{
		// the percent for the current segment of tVal
		// and the corresponding basis values
		for (size_t i=0; i<times.size()-2; ++i){
			if (times[i] <= tVal && tVal < times[i+1]){
				return i;
			}
		}
		return 0;
	}
}

ProgPairs Progression::getSplineOutput(double tVal, double mul) const{
	std::vector<const ProgPair* > sided;
	for (size_t i = 0; i < pairs.size(); ++i) {
		sided.push_back(&(pairs[i]));
	}
	return getRawSplineOutput(sided, tVal, mul);
}

ProgPairs Progression::getSplitSplineOutput(double tVal, double mul) const{
	std::vector<const ProgPair* > sided;
	bool gt = tVal >= 0.0;
	for (size_t i = 0; i < pairs.size(); ++i) {
		if (gt && pairs[i].second >= 0)
			sided.push_back(&(pairs[i]));
		else if (!gt && pairs[i].second <= 0)
			sided.push_back(&(pairs[i]));
	}
	return getRawSplineOutput(sided, tVal, mul);
}

ProgPairs Progression::getLinearOutput(double tVal, double mul) const{
	std::vector<const ProgPair* > sided;
	for (size_t i = 0; i < pairs.size(); ++i) {
		sided.push_back(&(pairs[i]));
	}
	return getRawLinearOutput(sided, tVal, mul);
}

ProgPairs Progression::getRawSplineOutput(const std::vector<const std::pair<Shape*, double>* > pairs, double tVal, double mul){
	if (
		(pairs.size() <= 2) ||
		(tVal < pairs[0]->second) && (tVal > pairs[pairs.size()-1]->second)
	){
		return getRawLinearOutput(pairs, tVal, mul);
	}

	std::vector<Shape*> shapes;
	std::vector<double> st;
	for (auto it = pairs.begin(); it != pairs.end(); ++it){
		shapes.push_back((*it)->first);
		st.push_back((*it)->second);
	}
	bool outside = false;
	size_t interval = getInterval(tVal, st, outside);
	ProgPairs out;

	double start = st[interval];
	double end = st[interval + 1];

	//# compute the catmull-rom basis multipliers
	double x = (tVal - start) / (end - start);
	if (outside) {
		// If I'm outside the range of the spline, then I linear interpolate along the implicit tangent
		if (interval == 0) {
			out.push_back(std::make_pair(shapes[0], mul * (1.0 - x)));
			out.push_back(std::make_pair(shapes[1], mul * x));
		}
		else {
			out.push_back(std::make_pair(shapes[shapes.size() - 1], mul * x));
			out.push_back(std::make_pair(shapes[shapes.size() - 2], mul * (1.0 - x)));
		}
	}
	else{
		double x2 = x*x;
		double x3 = x2*x;
		double v0 = (-0.5*x3 + 1.0*x2 - 0.5*x);
		double v1 = (1.5*x3 - 2.5*x2 + 1.0);
		double v2 = (-1.5*x3 + 2.0*x2 + 0.5*x);
		double v3 = (0.5*x3 - 0.5*x2);
		if (interval == 0) { // deal with input tangent
			out.push_back(std::make_pair(shapes[0], mul * (v1 + v0 + v0)));
			out.push_back(std::make_pair(shapes[1], mul * (v2 - v0)));
			out.push_back(std::make_pair(shapes[2], mul * (v3)));
		}
		else if (interval == st.size() - 2) { // deal with output tangent
			out.push_back(std::make_pair(shapes[shapes.size() - 3], mul * (v0)));
			out.push_back(std::make_pair(shapes[shapes.size() - 2], mul * (v1 - v3)));
			out.push_back(std::make_pair(shapes[shapes.size() - 1], mul * (v2 + v3 + v3)));
		}
		else {
			out.push_back(std::make_pair(shapes[interval - 1], mul * v0));
			out.push_back(std::make_pair(shapes[interval + 0], mul * v1));
			out.push_back(std::make_pair(shapes[interval + 1], mul * v2));
			out.push_back(std::make_pair(shapes[interval + 2], mul * v3));
		}
	}
	return out;
}

ProgPairs Progression::getRawLinearOutput(const std::vector<const std::pair<Shape*, double>* > pairs, double tVal, double mul){
	ProgPairs out;
	std::vector<double> times;
	if (pairs.size() < 2) return out;

	for (auto it=pairs.begin(); it!=pairs.end(); ++it){
		times.push_back((*it)->second);
	}
	bool outside;
	size_t idx = getInterval(tVal, times, outside);
	double u = (tVal - times[idx]) / (times[idx+1] - times[idx]);
    out.push_back(std::make_pair(pairs[idx]->first, mul * (1.0-u)));
	out.push_back(std::make_pair(pairs[idx+1]->first, mul * u));
	return out;
}

ProgPairs Progression::getOutput(double tVal, double mul) const{
	if (interp == ProgType::spline)
		return getSplineOutput(tVal, mul);
	else if (interp == ProgType::splitSpline)
		return getSplitSplineOutput(tVal, mul);
	else // if (interp == ProgType::linear)
		return getLinearOutput(tVal, mul);
}

bool Progression::parseJSONv1(const rapidjson::Value &val, size_t index, Simplex *simp){
	if (!val.IsArray()) return false;

	ProgPairs pairs;
	const rapidjson::Value &jindices = val[1];
	const rapidjson::Value &jweights = val[2];
	if (!jweights.IsArray() || !jindices.IsArray()) return false;

	rapidjson::SizeType j;
	for (j = 0; j<jindices.Size(); ++j){
		if (!jindices[j].IsInt()) return false;
		if (!jweights[j].IsNumber()) return false;
		size_t x = (size_t)jindices[j].GetInt();
		double y = (double)jweights[j].GetDouble();
		if (x >= simp->shapes.size()) return false;
		pairs.push_back(std::make_pair(&simp->shapes[x], y));
	}

	if (!val[0u].IsString()) return false;
	std::string name(val[0u].GetString());
	
	ProgType interp = ProgType::spline;

	if (val.Size() > 3){
		if (!val[3].IsString()) return false;
		std::string interpStr = val[3].GetString();
		if (interpStr == "linear") {
			interp = ProgType::linear;
		}
	}
	simp->progs.push_back(Progression(name, pairs, interp));
	return true;
}

bool Progression::parseJSONv2(const rapidjson::Value &val, size_t index, Simplex *simp){
	if (!val.IsObject()) return false;

	CHECK_JSON_STRING(nameIt, "name", val);
	CHECK_JSON_ARRAY(pairsIt, "pairs", val);
	CHECK_JSON_STRING(interpIt, "interp", val);

	std::string name(nameIt->value.GetString());

	std::string interpStr(interpIt->value.GetString());
	ProgType interp = ProgType::spline;
	if (interpStr == "linear")
		interp = ProgType::linear;
	else if (interpStr == "splitspline")
		interp = ProgType::splitSpline;

	ProgPairs pairs;

	auto &pairsVal = pairsIt->value;
	for (auto it = pairsVal.Begin(); it != pairsVal.End(); ++it){
		auto &ival = *it;
		if (!ival.IsArray()) return false;
		if (!ival[0].IsInt()) return false;
		if (!ival[1].IsDouble()) return false;

		size_t x = (size_t)ival[0].GetInt();
		double y = (double)ival[1].GetDouble();

		if (x >= simp->shapes.size()) return false;
		pairs.push_back(std::make_pair(&simp->shapes[x], y));
	}
	simp->progs.push_back(Progression(name, pairs, interp));
	return true;
}

bool Progression::parseJSONv3(const rapidjson::Value &val, size_t index, Simplex *simp){
	return parseJSONv2(val, index, simp);
}

File: src/python/simplexui/items/progression.py
# Copyright 2016, Blur Studio
#
# This file is part of Simplex.
#
# Simplex is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Simplex is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with Simplex.  If not, see <http://www.gnu.org/licenses/>.


# pylint:disable=missing-docstring,unused-argument,no-self-use
from __future__ import annotations
from ..utils import getNextName
from .accessor import SimplexAccessor
from .stack import stackable


from typing import Optional, Union, Any, TYPE_CHECKING

if TYPE_CHECKING:
    from .simplex import Simplex
    from .shape import Shape
    from .falloff import Falloff
    from .slider import Slider
    from .combo import Combo
    from .traversal import Traversal


class ProgPair(SimplexAccessor):
    classDepth: int = 9

    def __init__(self, simplex: Simplex, shape: Shape, value: float):
        super(ProgPair, self).__init__(simplex)
        self.shape: Shape = shape
        self._value: float = value
        self.prog: Optional[Progression] = None
        self.minValue: float = -1.0
        self.maxValue: float = 1.0
        if not shape.isRest and self not in self.shape.progPairs:
            self.shape.progPairs.append(self)

    @property
    def name(self) -> str:
        return self.shape.name

    @name.setter
    def name(self, value: str):
        self.shape.name = value

    def buildDefinition(
        self, simpDict: dict[str, Any], legacy: bool
    ) -> tuple[int, float]:
        idx = self.shape.buildDefinition(simpDict, legacy)
        return idx, self.value

    def __lt__(self, other: Any) -> bool:
        return self.value < other.value

    @property
    def value(self) -> float:
        return self._value

    @value.setter
    @stackable
    def value(self, val: float):
        from .slider import Slider

        self._value = val
        if self.prog is None:
            raise ValueError(
                "Can't set the value of a progPair that doesn't have a controller"
            )

        if isinstance(self.prog.controller, Slider):
            self.prog.controller.setRange()

    @stackable
    def delete(self):
        if self.prog is None:
            return

        ppairs = self.prog.pairs
        ridx = ppairs.index(self)
        pp = ppairs.pop(ridx)
        if not pp.shape.isRest:
            pp.shape.progPairs.remove(pp)
            if not pp.shape.progPairs:
                self.simplex.shapes.remove(pp.shape)
                self.DCC.deleteShape(pp.shape)


class Progression(SimplexAccessor):
    """A set of shapes to interpolate between

    A Progression is a collection of shape/value pairs, and an interpolation type.
    Progressions don't exist on their own, they are always part of a higher-level object
    like a Combo, Slider, or Traversal. The ProgPairs are always sorted by value

    Progressions should always have a shape at 0 (which is almost always the rest shape)
    and a shape at either -1 or 1.
    They can also have other shapes at any value between 0 and the extremes.

    Sliders give users direct control over the value that is passed to the progression.
    Combos and Traversals use input values to control their progressions.

    Progressions can use different interpolations.
    The simplest is 'linear', which blends in a straight line between shapes.
    The 'spline' interp uses Catmull-Rom spline values.
    The 'splitspline' builds separate Catmull-Rom splines for positive and negative values

    Parameters
    ----------
    name : str
        The name for the Progression. Usually just copies the name of its controller
    simplex : Simplex
        The Simplex system
    pairs : [ProgPair
        The ProgPairs that will make up this Progression.
        If None, the a default Rest at 0.0 pair will be created.
    interp : str
        The interpolation for this Progression. Defaults to 'spline'
    falloffs : [Falloff
        A list of Fallofs to apply to the progression
        Defaults to None

    Returns
    -------

    """

    classDepth: int = 8
    interpTypes: tuple[tuple[str, str], ...] = (
        ("Linear", "linear"),
        ("Spline", "spline"),
        ("Split Spline", "splitspline"),
    )

    def __init__(
        self,
        name: str,
        simplex: Simplex,
        pairs: Optional[list[ProgPair]] = None,
        interp: str = "spline",
        falloffs: Optional[list[Falloff]] = None,
    ):
        super(Progression, self).__init__(simplex)

        if self.simplex.restShape is None:
            raise RuntimeError("Simplex is not initialized with a rest shape")

        with self.stack.store(self):
            self._name: str = name
            self._interp: str = interp
            self.falloffs: list[Falloff] = falloffs or []
            self.controller: Optional[Union[Slider, Combo, Traversal]] = None

            self.pairs: list[ProgPair]
            if pairs is None:
                self.pairs = [ProgPair(self.simplex, self.simplex.restShape, 0.0)]
            else:
                self.pairs = pairs

            for pair in self.pairs:
                pair.prog = self

            for falloff in self.falloffs:
                falloff.children.append(self)
            self._buildIdx: Optional[int] = None

    @property
    def name(self) -> str:
        return self._name

    @name.setter
    def name(self, value: str):
        self._name = value

    @property
    def interp(self) -> str:
        """Get the interp for this Progression"""
        return self._interp

    @interp.setter
    @stackable
    def interp(self, value: str):
        """Set the interp for this Progression"""
        self._interp = value

    def getShapeIndex(self, shape: Shape) -> int:
        """Get the index of the given shape in this progression

        Parameters
        ----------
        shape : Shape
            The shape to get the index of

        Returns
        -------
        : int
            The index of the ProgPair that contains the given shape

        """
        for i, p in enumerate(self.pairs):
            if p.shape == shape:
                return i
        raise ValueError("Provided shape:{0} is not in the list".format(shape.name))

    def getShapes(self) -> list[Shape]:
        """Return the Shapes in this Progression

        Returns
        -------
        : type
            ([Shape, ....]): The shapes in the Progression

        """
        return [i.shape for i in self.pairs]

    def getValues(self) -> list[float]:
        """Return the values in this Progression

        Returns
        -------
        : type
            ([float, ....]): The values in the Progression

        """
        return [i.value for i in self.pairs]

    def getInsertIndex(self, tVal: float) -> int:
        """Get the index to insert a pair with value tVal

        Parameters
        ----------
        tVal : float
            The value to get the insertion index for

        Returns
        -------
        : int
            The insertion index

        """
        values = self.getValues()
        if not values:
            return 0
        elif tVal <= values[0]:
            return 0
        elif tVal >= values[-1]:
            return len(self.pairs)
        else:
            for i in range(1, len(values)):
                if values[i - 1] <= tVal < values[i]:
                    return i
        return 0

    def getShapeAtValue(self, val: float, tol=0.0001) -> Optional[Shape]:
        """Return the shape at the given value

        Parameters
        ----------
        val :
            float
        tol :
            float (Default value = 0.0001)

        Returns
        -------
        : type
            (Shape or None): The shape found with the given value, or None if nothing was found

        """
        for pp in self.pairs:
            if abs(pp.value - val) < tol:
                return pp.shape
        return None

    @classmethod
    def loadV2(cls, simplex: Simplex, data: dict[str, Any]) -> Progression:
        """Load the data from a version2 formatted json dictionary

        Parameters
        ----------
        simplex : Simplex
            The Simplex system that's being built
        data : dict
            The chunk of the json dict used to build this object

        Returns
        -------
        : Progression
            The specified Progression

        """
        name = data["name"]
        pairs = data["pairs"]
        interp = data.get("interp", "spline")
        foIdxs = data.get("falloffs", [])
        pairs = [ProgPair(simplex, simplex.shapes[s], v) for s, v in pairs]
        fos = [simplex.falloffs[i] for i in foIdxs]
        return cls(name, simplex, pairs=pairs, interp=interp, falloffs=fos)

    def buildDefinition(self, simpDict: dict[str, Any], legacy: bool) -> int:
        """Output a dictionary definition of this object

        Parameters
        ----------
        simpDict : dict
            The dictionary that is being built
        legacy : bool
            Whether to write out the legacy definition, or the newer one
        """
        if self._buildIdx is None:
            idxPairs = [pair.buildDefinition(simpDict, legacy) for pair in self.pairs]
            idxPairs.sort(key=lambda x: x[1])
            idxs, values = list(zip(*idxPairs))
            foIdxs = [f.buildDefinition(simpDict, legacy) for f in self.falloffs]
            self._buildIdx = len(simpDict["progressions"])
            if legacy:
                x = [self.name, idxs, values, self.interp, foIdxs]
                simpDict.setdefault("progressions", []).append(x)
            else:
                x = {
                    "name": self.name,
                    "pairs": idxPairs,
                    "interp": self.interp,
                    "falloffs": foIdxs,
                }
                simpDict.setdefault("progressions", []).append(x)
        return self._buildIdx

    def clearBuildIndex(self):
        """Clear the build index of this object

        The buildIndex is stored when building a definition dictionary
        that keeps track of its index for later referencin
        """
        self._buildIdx = None
        for pair in self.pairs:
            pair.shape.clearBuildIndex()
        for fo in self.falloffs:
            fo.clearBuildIndex()

    @stackable
    def moveShapeToProgression(self, shapePair: ProgPair):
        """Remove the shapePair from its current progression and set it in a new progression

        Parameters
        ----------
        shapePair : progPair
            The ProgPair to take
        """
        oldProg = shapePair.prog
        if oldProg is not None:
            oldProg.pairs.remove(shapePair)
        self.pairs.append(shapePair)
        shapePair.prog = self

    @stackable
    def setShapesValues(self, values: list[float]):
        """Set all the Shape's values

        Parameters
        ----------
        values : [float
            The values to set
        """
        from .slider import Slider

        for pp, val in zip(self.pairs, values):
            pp.value = val

        if isinstance(self.controller, Slider):
            self.controller.updateRange()

    def siblingRename(
        self, shape: Shape, newName: str, currentLinks: dict
    ) -> dict[type, dict[str, tuple[SimplexAccessor, int]]]:
        # This is part of the in-progress linked naming system
        # get name change
        return {}

    @stackable
    def addFalloff(self, falloff: Falloff):
        """Add a falloff to a slider's falloff list

        Parameters
        ----------
        falloff : Falloff
            The falloff to add

        Returns
        -------

        """
        if falloff not in self.falloffs:
            self.falloffs.append(falloff)
            falloff.children.append(self)
            self.DCC.addProgFalloff(self, falloff)

    @stackable
    def removeFalloff(self, falloff: Falloff):
        """Remove a falloff from a slider's falloff list

        Parameters
        ----------
        falloff : Falloff
            The falloff to remove

        Returns
        -------

        """
        if falloff in self.falloffs:
            self.falloffs.remove(falloff)
            falloff.children.remove(self)
            self.DCC.removeProgFalloff(self, falloff)

    @stackable
    def createShape(
        self, shapeName: Optional[str] = None, tVal: Optional[float] = None
    ) -> ProgPair:
        """Create a shape and add it to a progression

        Parameters
        ----------
        shapeName : str or None
            The name to give the shape
            If None, give it a default value
        tVal : float or None
            The value to give the new ProgPair
            if None, give it a "smart" default

        Returns
        -------
        : ProgPair
            The newly created ProgPair

        """
        from .slider import Slider

        pp, idx = self.newProgPair(shapeName, tVal)
        pp.prog = self
        self.pairs.insert(idx, pp)

        if isinstance(self.controller, Slider):
            self.controller.updateRange()

        return pp

    def newProgPair(
        self, shapeName: Optional[str] = None, tVal: Optional[float] = None
    ) -> tuple[ProgPair, int]:
        """Create a shape and DO NOT add it to a progression

        Parameters
        ----------
        shapeName : str or None
            The name to give the shape
            If None, give it a default value
        tVal : float or None
            The value to give the new ProgPair
            if None, give it a "smart" default

        Returns
        -------
        : ProgPair
            The newly created ProgPair
        : int
            The insertion index for this ProgPair into this Progression

        """
        from .shape import Shape

        if tVal is None:
            tVal = self.guessNextTVal()

        if shapeName is None:
            if self.controller is not None:
                if abs(tVal) == 1.0:
                    shapeName = self.controller.name
                else:
                    neg = "n" if tVal < 0.0 else ""
                    shapeName = "{0}_{1}{2}".format(
                        self.controller.name, neg, int(abs(tVal) * 100)
                    )

            currentNames = [i.name for i in self.simplex.shapes]
            shapeName = getNextName(shapeName, currentNames)

        idx = self.getInsertIndex(tVal)
        shape = Shape(shapeName, self.simplex)
        pp = ProgPair(self.simplex, shape, tVal)
        return pp, idx

    def guessNextTVal(self) -> float:
        """Given the current progression values, make an educated guess what's next.

        Returns
        -------
        : float
            The "smart" guess for the next tVal

        """
        # The question remains if negative or
        # intermediate values are more important
        # I think intermediate
        vals = [i.value for i in self.pairs]
        mnv = min(vals)
        mxv = max(vals)
        if mnv == 0.0 and mxv == 1.0:
            for c in [0.5, 0.25, 0.75, -1.0]:
                if c not in vals:
                    return c
        if mnv == -1.0 and mxv == 1.0:
            for c in [0.5, -0.5, 0.25, -0.25, 0.75, -0.75]:
                if c not in vals:
                    return c
        return 1.0

    @stackable
    def deleteShape(self, shape: Shape):
        """Delete a shape from the system and the DCC

        Parameters
        ----------
        shape : Shape
            The shape to delete
        """
        ridx = None
        for i, pp in enumerate(self.pairs):
            if pp.shape == shape:
                ridx = i
        if ridx is None:
            raise RuntimeError("Shape does not exist to remove")

        pp = self.pairs[ridx]
        self.pairs.pop(ridx)
        if not shape.isRest:
            self.simplex.shapes.remove(shape)
            self.DCC.deleteShape(shape)

    @stackable
    def delete(self):
        """Delete the Progression and all its Shapes"""
        for pp in self.pairs[:]:
            if pp.shape.isRest:
                continue
            self.simplex.shapes.remove(pp.shape)
            self.DCC.deleteShape(pp.shape)

    def getRange(self) -> tuple[float, float]:
        """Get the range for this Progression
        Returns
        -------
        : float
            The minimum value
        : float
            The maximum value

        """
        vals = [i.value for i in self.pairs]
        return min(vals), max(vals)

    def getExtremePairs(self) -> list[ProgPair]:
        """Get the ProgPairs where the value is -1 or 1

        Returns
        -------
        : [ProgPair, ...]
            ProgPairs whose values are -1 or 1
        """
        ret = []
        for pp in self.pairs:
            if abs(pp.value) != 1.0:
                continue
            ret.append(pp)
        return ret
